torch/csrc/jit/script/compiler.cpp

```diff
@@ -306,8 +323,15 @@ struct Environment {
             << value->kind() << " and " << name
             << " is not a first-class value.  Only reassignments to first-class values are allowed";
       }
-      if (!as_simple_value->type()->isSubtypeOf(
-              unshapedType(simple_parent->type()))) {
+
+      auto parent_type = unshapedType(simple_parent->type());
+      as_simple_value = tryConvertToType(
+          loc,
+          *b->owningGraph(),
+          parent_type,
+          as_simple_value,
+          /*allow_conversions=*/true);
+      if (!as_simple_value->type()->isSubtypeOf(parent_type)) {
```

Comment:
I think the only case where `tryConvertToType` will return a new type and that type will be a subtype of the parent type is with a None and an optional. Another option is to break out that part of the `tryConvertToType` function so this code will be more transparent.

Reply:
We also want variable assignment to follow all the same type conversion rules as schemas, so using tryConvertToType is correct here.